lib/exhal/transcoder.ex

```diff
@@ -79,19 +79,25 @@ defmodule ExHal.Transcoder do
   initial_params - the initial params with which the newly extracted info should
     merged.
   src_doc - the document to interpret
+  opts - options for use by modules adopting ExHal.ValueConverterWithOptions behaviour
   """
-  @callback decode!(%{}, ExHal.Document.t) :: %{}
   @callback decode!(ExHal.Document.t) :: %{}
+  @callback decode!(%{}, ExHal.Document.t) :: %{}
+  @callback decode!(ExHal.Document.t, [key: any]) :: %{}
+  @callback decode!(%{}, ExHal.Document.t, [key: any]) :: %{}
 
   @callbackdoc"""
   Returns an HAL version of params provided, combined with the initial doc.
 
   initial_doc - the initial document with which the newly encoded info should
     merged.
   src_params - the params to encoded into HAL
+  opts - options for use by modules adopting ExHal.ValueConverterWithOptions behaviour
   """
-  @callback encode!(Exhal.Document.t, %{}) :: ExHal.Document.t
   @callback encode!(%{}) :: ExHal.Document.t
+  @callback encode!(Exhal.Document.t, %{}) :: ExHal.Document.t
+  @callback encode!(%{}, [key: any]) :: ExHal.Document.t
+  @callback encode!(Exhal.Document.t, %{}, [key: any]) :: ExHal.Document.t
```

Comment:
I just realized there is a builtin type `keyword`. We should use that instead of the ad hoc definition i suggested earlier.